File: 061-080/challenge_073.py
'''
Given the head of a singly linked list, reverse it in-place.
'''

class SLLNode:
    def __init__(self, value):
        self.value = value
        self.next = None

class SLL:
    def __init__(self):
        self.head = None
        self._length = 0

    def __len__(self):
        return self._length

    def __repr__(self):
        out_str = "["
        node_ptr = self.head
        for i in range(0, self._length):
            out_str += str(node_ptr.value)
            if i < self._length - 1:
                out_str += "->"
            node_ptr = node_ptr.next
        out_str += "]"
        return out_str

    def addFront(self, value):
        new_node = SLLNode(value)

        if self.head:
            new_node.next = self.head
        self.head = new_node
        self._length += 1
        return self

    def addBack(self, value):
        new_node = SLLNode(value)
        self._length += 1

        if not self.head:
            self.head = new_node
            return self

        node_ptr = self.head
        while node_ptr.next:
            node_ptr = node_ptr.next
        node_ptr.next = new_node
        return self

    # Reverse Linked List In Place Function
    def reverse(self):
        if not self.head:
            return self
        if not self.head.next:
            return self

        left = self.head
        center = self.head.next
        left.next = None
        while center:
            right = center.next
            center.next = left
            left = center
            center = right
        self.head = left
        return self

sll = SLL()
sll.addFront(5).addFront(4).addFront(3).addFront(2).addFront(1)
print(sll)
sll.reverse()
print(sll)

File: js/061-080/challenge_073.js
/*
Given the head of a singly linked list, reverse it in-place.
*/

class SLLNode {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

class SLL {
    #length = 0;

    constructor() {
        this.head = null;
        for (let i = arguments.length; i > 0; i--) {
            this.addToFront(arguments[i-1]);
        }
    }

    print() {
        let outStr = "";
        let nodePtr = this.head;
        while (nodePtr != null) {
            outStr = outStr.concat(nodePtr.data.toString());
            if (nodePtr.next != null) {
                outStr = outStr.concat("->");
            }
            nodePtr = nodePtr.next;
        }
        console.log(outStr);
        return;
    }

    addToFront(data) {
        const newNode = new SLLNode(data);
        newNode.next = this.head;
        this.head = newNode;
        this.#length++;
        return;
    }

    addToBack(data) {
        const newNode = new SLLNode(data);
        this.#length++;

        if (this.head == null) {
            this.head = newNode;
            return;
        }

        let nodePtr = this.head;
        while (nodePtr.next != null) {
            nodePtr = nodePtr.next;
        }
        nodePtr.next = newNode;
        return;
    }

    reverse() {
        if (this.head == null || this.head.next == null) {
            return;
        }

        let prv = this.head;
        let cur = this.head.next;
        prv.next = null;
        while (cur != null) {
            let nxt = cur.next;
            cur.next = prv;
            prv = cur;
            cur = nxt;
        }
        this.head = prv;
        return;
    }
}

const sll = new SLL(1, 2, 3, 4, 5);
sll.print();
sll.reverse();
sll.print();